src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2251 @@
+                                          circpad_delay_t target_bin_us,
+                                          int use_usec);
+STATIC void circpad_machine_setup_tokens(circpad_machineinfo_t *mi);
+static inline circpad_purpose_mask_t circpad_circ_purpose_to_mask(uint8_t
+                                          circ_purpose);
+static inline circpad_circuit_state_t circpad_circuit_state(
+                                        origin_circuit_t *circ);
+static void circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx);
+static void circpad_setup_machine_on_circ(circuit_t *on_circ,
+                                          circpad_machine_t *machine);
+static double circpad_distribution_sample(circpad_distribution_t dist);
+
+/** Cached consensus params */
+static uint8_t circpad_per_hop_max_padding_percent;
+static uint16_t circpad_per_hop_allowed_cells;
+
+/* Machines for various usecases */
+static smartlist_t *origin_padding_machines = NULL;
+static smartlist_t *relay_padding_machines = NULL;
+
+/**
+ * Return the circpad_stat_t for the current state based on the
+ * mutable info.
+ */
+STATIC const circpad_state_t *
+circpad_machine_current_state(circpad_machineinfo_t *machine)
+{
+  switch (machine->current_state) {
+    case CIRCPAD_STATE_END:
+      return NULL;
+
+    case CIRCPAD_STATE_START:
+      return &CIRCPAD_GET_MACHINE(machine)->start;
+
+    case CIRCPAD_STATE_BURST:
+      return &CIRCPAD_GET_MACHINE(machine)->burst;
+
+    case CIRCPAD_STATE_GAP:
+      return &CIRCPAD_GET_MACHINE(machine)->gap;
+  }
+
+  log_fn(LOG_WARN,LD_CIRC,
+         "Invalid circuit padding state %d",
+         machine->current_state);
+  //tor_fragile_assert();
+
+  return NULL;
+}
+
+/**
+ * Calculate the lower bound of a histogram bin. The upper bound
+ * is obtained by calling this function with bin+1, and subtracting 1.
+ *
+ * The 0th bin has a special value -- it only represents start_usec.
+ * This is so we can specify a probability on 0-delay values.
+ *
+ * After bin 0, bins are exponentially spaced, so that each subsequent
+ * bin is twice as large as the previous. This is done so that higher
+ * time resolution is given to lower time values.
+ *
+ * The infinity bin is a the last bin in the array (histogram_len-1).
+ * It has a usec value of CIRCPAD_DELAY_INFINITE (UINT32_MAX).
+ */
+STATIC circpad_delay_t
+circpad_histogram_bin_to_usec(circpad_machineinfo_t *mi, int bin)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  circpad_delay_t start_usec;
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (bin >= state->histogram_len-1)
+    return CIRCPAD_DELAY_INFINITE;
+
+  if (bin == 0)
+    return start_usec;
+
+  if (bin == 1)
+    return start_usec+1;
+
+  return start_usec
+      + (state->range_sec*USEC_PER_SEC)/(1<<(state->histogram_len-bin+1));
```

Comment:
Hrmm. A bunch of 0 sized bins.

We should probably have a circpad_machine_validate() function that checks for obvious signs of machine degeneracy like this while initializing machines. We might be able to wait for the serialization code to be written before we do that, since we only have one hardcoded machine right now that is presumed to be valid.

Reply:
On a related note: this line is very complex, it contains 14-16 elements: 8 values and 6-8 operators (and 3 sets of brackets). Most humans are good with 3-7 things in their head at once.

For maintainability, please split it into at least 3 lines, using temporary variables. If you make your variables const, most compilers will eliminate the extra operations. (Some will do it Ben if they're not const.)

Reply:
5fa69f86ae6c4843599e2ad994fffb814e26aae4